RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -76,75 +78,91 @@ public static Function<Path, LuceneDocumentsReader> getFactory(boolean softDelet
      *    present in the Lucene Index.  This wrapper will filter out documents that are marked as "soft deleted" in the
      *    Lucene Index.
 
-    */
+     */
     public Flux<Document> readDocuments() {
-        return Flux.using(() -> wrapReader(DirectoryReader.open(FSDirectory.open(indexDirectoryPath)), softDeletesPossible, softDeletesField), reader -> {
-            log.atInfo().log(reader.maxDoc() + " documents found in the current Lucene index");
-
-            return Flux.fromIterable(reader.leaves()) // Iterate over each segment
-                .flatMap(leaf -> {
-                    LeafReader leafReader = leaf.reader();
-                    Bits liveDocs = leafReader.getLiveDocs(); // Get live docs bits
-
-                    return Flux.range(0, leafReader.maxDoc())
-                        .handle((docId, sink) -> {
-                            // Check if the document is live (not hard deleted/updated)
-                            boolean isLive = liveDocs == null || liveDocs.get(docId);
-                            Document doc = getDocument(leafReader, docId, isLive);
-
-                            if (doc != null) { // Skip malformed docs
-                                sink.next(doc);
-                            }
-                        });
-                })
-                .cast(Document.class);
-        }, reader -> { // Close the DirectoryReader when done
-            try {
-                reader.close();
-            } catch (IOException e) {
-                log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
-                throw Lombok.sneakyThrow(e);
-            }
+        return Flux.using(
+            () -> wrapReader(DirectoryReader.open(FSDirectory.open(indexDirectoryPath)), softDeletesPossible, softDeletesField),
+            this::readDocsByLeavesInParallel,
+            reader -> {
+                try {
+                    reader.close();
+                } catch (IOException e) {
+                    log.atError().setMessage("Failed to close DirectoryReader").setCause(e).log();
+                    throw Lombok.sneakyThrow(e);
+                }
         });
     }
 
+    Publisher<Document> readDocsByLeavesInParallel(DirectoryReader reader) {
+        var segmentsToReadAtOnce = 5; // Arbitrary value
+        var maxDocumentsToReadAtOnce = 100; // Arbitrary value
```

Comment:
I'm OK if these values give the throughput we are looking for - how does this all hold up with the latest perf numbers?